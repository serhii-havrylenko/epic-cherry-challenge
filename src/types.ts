export interface BaseOperationDefinition {
  id: string;
  parent: string | null;
}

export interface ConstantOperationDefinition {
  type: 'constant';
  value: boolean;
}

export interface GenericOperationDefinition {
  type: 'select' | 'argument' | 'or' | 'and';
  value: string | null;
}

export type OperationDefinition = BaseOperationDefinition &
  (ConstantOperationDefinition | GenericOperationDefinition);

export type Operations = Record<string, OperationDefinition>;

export interface VariableDefinition {
  id: string;
  name: string;
  value: boolean;
  error?: boolean;
}

export interface State {
  variables: VariableDefinition[];
  operations: Operations;
  rootNodeID: string;
}

export interface AddVariableAction {
  type: 'addVariable';
  payload: string;
}

export interface ChangeVariableNameAction {
  type: 'changeVariableName';
  payload: {
    id: string;
    name: string;
  };
}

export interface ChangeVariableValueAction {
  type: 'changeVariableValue';
  payload: {
    id: string;
    value: boolean;
  };
}

export interface SelectOperationAction {
  type: 'selectOperation';
  payload: {
    type: OperationDefinition['type'];
    id: string;
    value: string | null;
  };
}

export interface SetOperationValueAction {
  type: 'setOperationValue';
  payload: {
    id: string;
  } & (ConstantOperationDefinition | GenericOperationDefinition);
}

export interface DeleteOperationAction {
  type: 'deleteOperation';
  payload: string;
}

export interface AddOperationAction {
  type: 'addOperation';
  payload: {
    id: string;
    parent: string;
  };
}

export type Actions =
  | AddVariableAction
  | ChangeVariableNameAction
  | ChangeVariableValueAction
  | SelectOperationAction
  | SetOperationValueAction
  | DeleteOperationAction
  | AddOperationAction;

export type UseSelectorCallback<R = unknown> = (state: State) => R;
export type UseSelectorType = <R = unknown>(callback: UseSelectorCallback<R>) => R;
