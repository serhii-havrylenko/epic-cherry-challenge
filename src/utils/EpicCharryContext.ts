/* eslint-disable @typescript-eslint/no-unnecessary-type-constraint, @typescript-eslint/no-empty-function */
import { createContext, useContext } from 'react';
import { Actions, UseSelectorCallback, UseSelectorType } from '../types';

export const defaultGetValue: UseSelectorType = <R extends unknown>(
  callback: UseSelectorCallback<R>,
): R => {
  return undefined as R;
};

export const EpicCherryContext = createContext<{
  dispatch: React.Dispatch<Actions>;
  getValue: UseSelectorType;
}>({
  dispatch: () => {},
  getValue: defaultGetValue,
});

export const useSelector = <R extends unknown>(callback: UseSelectorCallback<R>) => {
  const { getValue } = useContext(EpicCherryContext);
  return getValue(callback);
};

export const useDispatch = () => {
  const { dispatch } = useContext(EpicCherryContext);
  return dispatch;
};
