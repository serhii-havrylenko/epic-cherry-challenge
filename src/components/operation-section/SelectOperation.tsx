import { getDefinedVariables } from '../../selectors';
import { OperationDefinition } from '../../types';
import { useSelector } from '../../utils';

export const SelectOperation = ({
  onValueSelect,
}: {
  onValueSelect: (type: OperationDefinition['type']) => void;
}) => {
  const variables = useSelector(getDefinedVariables);

  return (
    <select
      value="select"
      onChange={(e) => onValueSelect(e.target.value as OperationDefinition['type'])}
    >
      <option value="select" disabled>
        select
      </option>
      <option value="constant">constant</option>
      <option value="argument" disabled={variables.length === 0}>
        argument
      </option>
      <option value="and">and</option>
      <option value="or">or</option>
    </select>
  );
};
