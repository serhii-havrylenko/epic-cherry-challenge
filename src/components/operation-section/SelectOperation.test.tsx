import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { State } from '../../types';
import { EpicCherryContext } from '../../utils';
import { SelectOperation } from './SelectOperation';

test('renders correctly', () => {
  const state: State = {
    variables: [
      {
        id: 'var1',
        name: 'Var 1',
        value: true,
      },
    ],
    operations: {},
    rootNodeID: '1234',
  };
  const dispatchMock = jest.fn();

  const { container } = render(
    <EpicCherryContext.Provider
      value={{ getValue: (callback) => callback(state), dispatch: dispatchMock }}
    >
      <SelectOperation onValueSelect={jest.fn} />
    </EpicCherryContext.Provider>,
  );

  expect(container).toMatchSnapshot();
});

test('renders correctly and calls callback with the selected value', () => {
  const state: State = {
    variables: [
      {
        id: 'var1',
        name: 'Var 1',
        value: true,
      },
    ],
    operations: {},
    rootNodeID: '1234',
  };
  const dispatchMock = jest.fn();
  const onValueSelectMock = jest.fn();

  render(
    <EpicCherryContext.Provider
      value={{ getValue: (callback) => callback(state), dispatch: dispatchMock }}
    >
      <SelectOperation onValueSelect={onValueSelectMock} />
    </EpicCherryContext.Provider>,
  );

  userEvent.selectOptions(
    screen.getByRole('combobox'),
    screen.getByRole('option', { name: 'constant' }),
  );

  expect(onValueSelectMock).toHaveBeenCalledTimes(1);
  expect(onValueSelectMock).toHaveBeenCalledWith('constant');
});
