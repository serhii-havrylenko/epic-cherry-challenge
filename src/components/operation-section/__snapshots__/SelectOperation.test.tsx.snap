// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <select>
    <option
      disabled=""
      value="select"
    >
      select
    </option>
    <option
      value="constant"
    >
      constant
    </option>
    <option
      value="argument"
    >
      argument
    </option>
    <option
      value="and"
    >
      and
    </option>
    <option
      value="or"
    >
      or
    </option>
  </select>
</div>
`;
